src/java/org/apache/cassandra/locator/DynamicEndpointSnitchEMA.java

```diff
@@ -0,0 +1,181 @@
+        // arrive at sort of a 'badness percentage' for them. First, find the worst for each:
+        HashMap<InetAddressAndPort, Double> newScores = new HashMap<>(samples.size());
+        Optional<Double> maxObservedAvgLatency = samples.values().stream()
+                                                        .map(ExponentialMovingAverage::getAvg)
+                                                        .max(Double::compare);
+
+        final double maxAvgLatency = maxObservedAvgLatency.isPresent() ? maxObservedAvgLatency.get() : 1;
+
+        // now make another pass to do the weighting based on the maximums we found before
+        for (Map.Entry<InetAddressAndPort, AnnotatedEMA> entry : samples.entrySet())
+        {
+            // Samples may have changed but rather than creating garbage by copying samples we just ensure
+            // that all scores are less than 1.0
+            double addrAvg = entry.getValue().getAvg();
+            double score = addrAvg / Math.max(addrAvg, maxAvgLatency);
+            // finally, add the severity without any weighting, since hosts scale this relative to their own load
+            // and the size of the task causing the severity.
+            // "Severity" is basically a measure of compaction activity (CASSANDRA-3722).
+            if (USE_SEVERITY)
+                score += getSeverity(entry.getKey());
+            // lowest score (least amount of badness) wins.
+            newScores.put(entry.getKey(), score);
+        }
+        return newScores;
+    }
+
+    @Override
+    public List<Double> dumpTimings(String hostname) throws UnknownHostException
+    {
+        InetAddressAndPort host = InetAddressAndPort.getByName(hostname);
+        ArrayList<Double> timings = new ArrayList<>();
+        ExponentialMovingAverage avg = samples.get(host);
+        if (avg != null)
+        {
+            timings.add(avg.getAvg());
+        }
+        return timings;
+    }
+
+    @Override
+    protected void updateLatencyProbeSequence(List<InetAddressAndPort> probeSequence)
+    {
+        samples.keySet().retainAll(Gossiper.instance.getLiveMembers());
+
+        for (Map.Entry<InetAddressAndPort, AnnotatedEMA> entry: samples.entrySet())
+        {
+            // We only send latency probes to nodes that we may plausibly talk to (requested is true)
+            // but we have not talked to since the last reset of this information (recentlyMeasured is false)
+            if (entry.getValue().recentlyRequested && !entry.getValue().recentlyMeasured)
+            {
+                probeSequence.add(entry.getKey());
+            }
+
+            entry.getValue().recentlyMeasured = false;
+            entry.getValue().recentlyRequested = false;
+        }
+    }
+
+    @Override
+    protected void markRequested(InetAddressAndPort address)
+    {
+        AnnotatedEMA ema = samples.get(address);
+        if (ema != null)
+            ema.recentlyRequested = true;
```

Comment:
When we set this boolean we also want to check if it's already set before setting it.

Reply:
I was curious about this if a load before the set would be faster or slower than just a set. I guess I decided that the performance difference was probably so negligible I should just go for most readable. I don't care though heh, changing with the flag update.

Reply:
It's about scalability more than performance. Every time you write to the cache line it has to be acquired in an exclusive state to write and it's invalidated in caches everywhere else. For volatile fields that are set to the same value repeatedly at high frequency it can be better to not do that.

It's more about avoiding death by a thousand cuts WRT to performance. If it were really really not clear I would say don't do it, but it's a simple enough pattern for setting a boolean.